packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-fields-transformer.ts

```diff
@@ -0,0 +1,40 @@
+import {
+  TransformerContextProvider,
+  TransformerPrepareStepContextProvider,
+  TransformerTransformSchemaStepContextProvider,
+} from '@aws-amplify/graphql-transformer-interfaces';
+import { updateTableForConnection } from '../resolvers';
+import { HasManyDirectiveConfiguration } from '../types';
+import { registerHasManyForeignKeyMappings, getRelatedTypeIndex, ensureFieldsArray, getFieldsNodes } from '../utils';
+import { DataSourceBasedDirectiveTransformer } from '../data-source-based-directive-transformer';
+
+export class HasManyDirectiveDDBFieldsTransformer implements DataSourceBasedDirectiveTransformer<HasManyDirectiveConfiguration> {
+  dbType: 'DYNAMODB';
```

Comment:
as elsewhere, please use static type for this

Reply:
You're referring to `ModelDataSourceStrategyDbType`, correct? I'd prefer not to do that here because it removes an existing compiler-enforced invariant used when selecting the correct `DataSourceBasedDirectiveTransformer`.

Here's an example to demonstrate my concerns. 
```ts
export const getHasManyDirectiveTransformer = (
  dbType: ModelDataSourceStrategyDbType,
): DataSourceBasedDirectiveTransformer<HasManyDirectiveConfiguration> => {
  switch (dbType) {
    case 'MYSQL':
    case 'POSTGRES':
      return new HasManyDirectiveDDBFieldsTransformer(dbType); // ❌  Argument of type 'ModelDataSourceStrategySqlDbType' is not assignable to parameter of type '"DYNAMODB"'.
    case 'DYNAMODB':
      return new HasManyDirectiveDDBFieldsTransformer(dbType);
  }
};
```
This is currently an error, which is what we want; by moving to `ModelDataSourceStrategyDbType`, it's no longer an error.

Is there a best of both worlds scenario here that I'm missing?

Reply:
I'm already a seeing consistency error, which manifests as a linting warning: 

![image](https://github.com/aws-amplify/amplify-category-api/assets/227982/672fe5d1-34b6-4bd9-b31a-2781e88e3778)

which is fixed with a default. I'm unclear why TS isn't able to infer exhaustiveness, since the types are defined such that there are only three possible string values. 

We can chat with other folks on the team to see if they have any ideas--in particular, the runtime folks are probably good to check with. But if we aren't able to come up with a solution that resolves the warnings, I'd prefer to go with the constants rather than hardcoded strings for value comparisons.

Reply:
By the way, my comment above for the usage in [has-many-directive-ddb-fields-transformer.ts](https://github.com/aws-amplify/amplify-category-api/pull/2330/files/d579ce982146e3d3ec6039422ab56ac294266c05#diff-9e2644d489b13192d40d6d86918daa01673645d7440a1bf15b13feeeac389d26) is incorrect -- that's a type, not a value. A constant isn't appropriate for that, and we don't have a type defined for the Dynamo DB data type, so the bare string is the right way to enforce the restriction you're after.

(We could consider adding a `type ModelDataSourceStrategyDynamoDbType = 'DYNAMODB'` and refactoring `type ModelDataSourceStrategyDbType = ModelDataSourceStrategyDynamoDbType | ModelDataSourceStrategySqlDbType` but I'm not sure whether that actually worth it.)

Reply:
This was nagging at me, so for grins I put this into [a TS 4.5.5 playground](https://www.typescriptlang.org/play?target=5&module=1&ts=4.5.5#code/C4TwDgpgBAsg9gEwgGwCIENjoMpwK4BOAxhNsAZhAOYioBGAKuNALxQDkqAmgHICCMAPKoAQuygAfWIhQYsuQiTIVg1ENgCOaRswDcAKFCRpSNJhz5ipcpRqbtTY23Ywu2AIoAZcVPYAFQWwGAHEAJQBRbHYDIjgAOwBnYChUUQB9dIYuP3CALhNZcwUrZVtaHScObn4hUWj9WMTk1w9PDJE0rJz8+FM5C0VrFTV7ekdWDhavesakqACgsMj2zuy8grN5SyUbVTstMeYoZwWQiKiDBvi5hIB3AEtgIgALBggkz0eICmRjqAAKBAVCA9GSbAYlXZqQ6QACU+TieAAtnRvscAHxQADe+igUDujxeAKB41h2NxeKgRHQCWgnF4AmEYlyFMpUAIEGAhDiUAAjAY2dTaZM3NMWWy8RyuQQeQAmAWUoV005LKLiiVS7lQADMCoAvvoDbM4MgIAA6ZBwKj-dgEp6vd7AT6qH7sAA0+Ie9reHy+Pxt1UZdVhsMus2SdpePuAAGFrlg4sk2IDgaC+kVtkMyjCIPCoIiUWiWJicXjI89icCyaXFTToKkOpk1uq2ZqZXyFXilbBRW0m91WZS23LO1S6-NAmdlv28oPJZytbqKQajdcTebLdbbV6o464010In3Z7CQ6kvukofgAGGbUxCGDEA) and it works as expected with either strategy--constant or literal.

My best guess at this point is that we have a TSConfig option that is causing this to choke.

Reply:
> (We could consider adding a type ModelDataSourceStrategyDynamoDbType = 'DYNAMODB' and refactoring type ModelDataSourceStrategyDbType = ModelDataSourceStrategyDynamoDbType | ModelDataSourceStrategySqlDbType but I'm not sure whether that actually worth it.)

I actually tried this and ended up hitting a blocker.
```ts
src/types/model-datasource-strategies.ts:7:14 - error TS2322: Type '"DYNAMODB"' is not assignable to type '"DNAMODB"'.

7 export const DDB_DB_TYPE: ModelDataSourcesStrategyDynamoDbType = 'DYNAMODB';
```

Here's the patch in case anything jumps out at you.
```diff
diff --git a/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-ddb-fields-transformer.ts b/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-ddb-fields-transformer.ts
index 80c7d9a67..fa6b2b932 100644
--- a/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-ddb-fields-transformer.ts
+++ b/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-ddb-fields-transformer.ts
@@ -1,4 +1,5 @@
 import {
+  ModelDataSourcesStrategyDynamoDbType,
   TransformerContextProvider,
   TransformerPrepareStepContextProvider,
   TransformerTransformSchemaStepContextProvider,
@@ -14,8 +15,8 @@ import { DataSourceBasedDirectiveTransformer } from '../data-source-based-direct
  * This should not be used for `@belongsTo(references: [String!])` definitions.
  */
 export class BelongsToDirectiveDDBFieldsTransformer implements DataSourceBasedDirectiveTransformer<BelongsToDirectiveConfiguration> {
-  dbType: 'DYNAMODB';
-  constructor(dbType: 'DYNAMODB') {
+  dbType: ModelDataSourcesStrategyDynamoDbType;
+  constructor(dbType: ModelDataSourcesStrategyDynamoDbType) {
     this.dbType = dbType;
   }
 
diff --git a/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-transformer-factory.ts b/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-transformer-factory.ts
index f1e47cb23..47493e203 100644
--- a/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-transformer-factory.ts
+++ b/packages/amplify-graphql-relational-transformer/src/belongs-to/belongs-to-directive-transformer-factory.ts
@@ -3,15 +3,16 @@ import { BelongsToDirectiveConfiguration } from '../types';
 import { DataSourceBasedDirectiveTransformer } from '../data-source-based-directive-transformer';
 import { BelongsToDirectiveSQLTransformer } from './belongs-to-directive-sql-transformer';
 import { BelongsToDirectiveDDBFieldsTransformer } from './belongs-to-directive-ddb-fields-transformer';
+import { DDB_DB_TYPE, MYSQL_DB_TYPE, POSTGRES_DB_TYPE } from '@aws-amplify/graphql-transformer-core';
 
 export const getBelongsToDirectiveTransformer = (
   dbType: ModelDataSourceStrategyDbType,
 ): DataSourceBasedDirectiveTransformer<BelongsToDirectiveConfiguration> => {
   switch (dbType) {
-    case 'MYSQL':
-    case 'POSTGRES':
+    case MYSQL_DB_TYPE:
+    case POSTGRES_DB_TYPE:
       return new BelongsToDirectiveSQLTransformer(dbType);
-    case 'DYNAMODB':
+    case DDB_DB_TYPE:
       return new BelongsToDirectiveDDBFieldsTransformer(dbType);
   }
 };
diff --git a/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-fields-transformer.ts b/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-fields-transformer.ts
index 2055f797a..af82a8415 100644
--- a/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-fields-transformer.ts
+++ b/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-fields-transformer.ts
@@ -1,4 +1,5 @@
 import {
+  ModelDataSourcesStrategyDynamoDbType,
   TransformerContextProvider,
   TransformerPrepareStepContextProvider,
   TransformerTransformSchemaStepContextProvider,
@@ -15,8 +16,8 @@ import { DataSourceBasedDirectiveTransformer } from '../data-source-based-direct
  * This should not be used for `@hasMany(references: [String!])` definitions.
  */
 export class HasManyDirectiveDDBFieldsTransformer implements DataSourceBasedDirectiveTransformer<HasManyDirectiveConfiguration> {
-  dbType: 'DYNAMODB';
-  constructor(dbType: 'DYNAMODB') {
+  dbType: ModelDataSourcesStrategyDynamoDbType;
+  constructor(dbType: ModelDataSourcesStrategyDynamoDbType) {
     this.dbType = dbType;
   }
 
diff --git a/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-transformer-factory.ts b/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-transformer-factory.ts
index 328243cdb..7181e109b 100644
--- a/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-transformer-factory.ts
+++ b/packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-transformer-factory.ts
@@ -3,15 +3,16 @@ import { HasManyDirectiveConfiguration } from '../types';
 import { DataSourceBasedDirectiveTransformer } from '../data-source-based-directive-transformer';
 import { HasManyDirectiveDDBFieldsTransformer } from './has-many-directive-ddb-fields-transformer';
 import { HasManyDirectiveSQLTransformer } from './has-many-directive-sql-transformer';
+import { DDB_DB_TYPE, MYSQL_DB_TYPE, POSTGRES_DB_TYPE } from '@aws-amplify/graphql-transformer-core';
 
 export const getHasManyDirectiveTransformer = (
   dbType: ModelDataSourceStrategyDbType,
 ): DataSourceBasedDirectiveTransformer<HasManyDirectiveConfiguration> => {
   switch (dbType) {
-    case 'MYSQL':
-    case 'POSTGRES':
+    case MYSQL_DB_TYPE:
+    case POSTGRES_DB_TYPE:
       return new HasManyDirectiveSQLTransformer(dbType);
-    case 'DYNAMODB':
-      return new HasManyDirectiveDDBFieldsTransformer(dbType);
+      case DDB_DB_TYPE:
+        return new HasManyDirectiveDDBFieldsTransformer(dbType);
   }
 };
diff --git a/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-ddb-fields-transformer.ts b/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-ddb-fields-transformer.ts
index e0436a52c..219d10a73 100644
--- a/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-ddb-fields-transformer.ts
+++ b/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-ddb-fields-transformer.ts
@@ -1,4 +1,5 @@
 import {
+  ModelDataSourcesStrategyDynamoDbType,
   TransformerContextProvider,
   TransformerPrepareStepContextProvider,
   TransformerTransformSchemaStepContextProvider,
@@ -14,8 +15,8 @@ import { DataSourceBasedDirectiveTransformer } from '../data-source-based-direct
  * This should not be used for `@hasOne(references: [String!])` definitions.
  */
 export class HasOneDirectiveDDBFieldsTransformer implements DataSourceBasedDirectiveTransformer<HasOneDirectiveConfiguration> {
-  dbType: 'DYNAMODB';
-  constructor(dbType: 'DYNAMODB') {
+  dbType: ModelDataSourcesStrategyDynamoDbType;
+  constructor(dbType: ModelDataSourcesStrategyDynamoDbType) {
     this.dbType = dbType;
   }
 
diff --git a/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-transformer-factory.ts b/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-transformer-factory.ts
index b83e17b57..7ac49ebd2 100644
--- a/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-transformer-factory.ts
+++ b/packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-transformer-factory.ts
@@ -3,15 +3,16 @@ import { HasOneDirectiveConfiguration } from '../types';
 import { DataSourceBasedDirectiveTransformer } from '../data-source-based-directive-transformer';
 import { HasOneDirectiveDDBFieldsTransformer } from './has-one-directive-ddb-fields-transformer';
 import { HasOneDirectiveSQLTransformer } from './has-one-directive-sql-transformer';
+import { DDB_DB_TYPE, MYSQL_DB_TYPE, POSTGRES_DB_TYPE } from '@aws-amplify/graphql-transformer-core';
 
 export const getHasOneDirectiveTransformer = (
   dbType: ModelDataSourceStrategyDbType,
 ): DataSourceBasedDirectiveTransformer<HasOneDirectiveConfiguration> => {
   switch (dbType) {
-    case 'MYSQL':
-    case 'POSTGRES':
+    case MYSQL_DB_TYPE:
+    case POSTGRES_DB_TYPE:
       return new HasOneDirectiveSQLTransformer(dbType);
-    case 'DYNAMODB':
+    case DDB_DB_TYPE:
       return new HasOneDirectiveDDBFieldsTransformer(dbType);
   }
 };
diff --git a/packages/amplify-graphql-transformer-core/src/types/model-datasource-strategies.ts b/packages/amplify-graphql-transformer-core/src/types/model-datasource-strategies.ts
index 31d986120..591694b97 100644
--- a/packages/amplify-graphql-transformer-core/src/types/model-datasource-strategies.ts
+++ b/packages/amplify-graphql-transformer-core/src/types/model-datasource-strategies.ts
@@ -1,10 +1,10 @@
 import {
   ModelDataSourceStrategy,
-  ModelDataSourceStrategyDbType,
+  ModelDataSourcesStrategyDynamoDbType,
   ModelDataSourceStrategySqlDbType,
 } from '@aws-amplify/graphql-transformer-interfaces';
 
-export const DDB_DB_TYPE: ModelDataSourceStrategyDbType = 'DYNAMODB';
+export const DDB_DB_TYPE: ModelDataSourcesStrategyDynamoDbType = 'DYNAMODB';
 
 export const MYSQL_DB_TYPE: ModelDataSourceStrategySqlDbType = 'MYSQL';
 
diff --git a/packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts b/packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts
index 5e460a0b7..6bbe22e93 100644
--- a/packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts
+++ b/packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts
@@ -5,6 +5,7 @@ import {
   ModelDataSourceStrategy,
   ModelDataSourceStrategyDbType,
   ModelDataSourceStrategySqlDbType,
+  ModelDataSourcesStrategyDynamoDbType,
   SQLLambdaModelDataSourceStrategy,
 } from '@aws-amplify/graphql-transformer-interfaces';
 import { DDB_DB_TYPE, ImportedRDSType, MYSQL_DB_TYPE, POSTGRES_DB_TYPE } from '../types';
@@ -103,8 +104,8 @@ export const isDynamoDbModel = (ctx: DataSourceStrategiesProvider, typename: str
 /**
  * Type predicate that returns true if `dbType` is the DynamoDB database type
  */
-export const isDynamoDbType = (dbType: ModelDataSourceStrategyDbType): dbType is 'DYNAMODB' => {
-  return dbType === 'DYNAMODB';
+export const isDynamoDbType = (dbType: ModelDataSourceStrategyDbType): dbType is ModelDataSourcesStrategyDynamoDbType => {
+  return dbType === DDB_DB_TYPE;
 };
 
 /**
diff --git a/packages/amplify-graphql-transformer-interfaces/src/model-datasource/types.ts b/packages/amplify-graphql-transformer-interfaces/src/model-datasource/types.ts
index ef0a20019..aa5a8c5ec 100644
--- a/packages/amplify-graphql-transformer-interfaces/src/model-datasource/types.ts
+++ b/packages/amplify-graphql-transformer-interfaces/src/model-datasource/types.ts
@@ -18,7 +18,12 @@ export interface ModelDataSourceStrategyBase {
 /**
  * All supported database types that can be used to resolve models.
  */
-export type ModelDataSourceStrategyDbType = 'DYNAMODB' | ModelDataSourceStrategySqlDbType;
+export type ModelDataSourceStrategyDbType = ModelDataSourcesStrategyDynamoDbType | ModelDataSourceStrategySqlDbType;
+
+/**
+ * DynamoDB supported database type that can be used to resovle models.
+ */
+export type ModelDataSourcesStrategyDynamoDbType = 'DNAMODB'
 
 /**
  * All supported SQL database types that can be used to resolve models.
@@ -29,7 +34,7 @@ export type ModelDataSourceStrategySqlDbType = 'MYSQL' | 'POSTGRES';
  * Use default CloudFormation type 'AWS::DynamoDB::Table' to provision table.
  */
 export interface DefaultDynamoDbModelDataSourceStrategy extends ModelDataSourceStrategyBase {
-  readonly dbType: 'DYNAMODB';
+  readonly dbType: ModelDataSourcesStrategyDynamoDbType;
   readonly provisionStrategy: 'DEFAULT';
 }
 
@@ -37,7 +42,7 @@ export interface DefaultDynamoDbModelDataSourceStrategy extends ModelDataSourceS
  * Use custom resource type 'Custom::AmplifyDynamoDBTable' to provision table.
  */
 export interface AmplifyDynamoDbModelDataSourceStrategy extends ModelDataSourceStrategyBase {
-  readonly dbType: 'DYNAMODB';
+  readonly dbType: ModelDataSourcesStrategyDynamoDbType;
   readonly provisionStrategy: 'AMPLIFY_TABLE';
 }
```